python/delta-kernel-python/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+        ScanBuilder(Some(sb))
+    }
+
+    fn build(&mut self) -> DeltaPyResult<Scan> {
+        let scan = self.0.take().unwrap().build()?;
+        Ok(Scan(scan))
+    }
+}
+
+#[pyclass]
+struct Scan(delta_kernel::scan::Scan);
+
+#[pymethods]
+impl Scan {
+    fn execute(
+        &self,
+        engine_interface: &PythonInterface,
+    ) -> DeltaPyResult<PyArrowType<Box<dyn RecordBatchReader + Send>>> {
+        let result_schema: SchemaRef =
+            Arc::new(self.0.schema().as_ref().try_into().map_err(|e| {
+                delta_kernel::Error::Generic(format!("Could not get result schema: {e}"))
+            })?);
+        let results = self.0.execute(engine_interface.0.as_ref())?;
+        let record_batch_iter = RecordBatchIterator::new(
+            results.into_iter().map(|res| {
+                let data = res
+                    .raw_data
+                    .map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
+                let record_batch: RecordBatch = data
+                    .into_any()
+                    .downcast::<ArrowEngineData>()
+                    .map_err(|_| {
+                        ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string())
+                    })?
+                    .into();
+                if let Some(mut mask) = res.mask {
+                    let extra_rows = record_batch.num_rows() - mask.len();
+                    if extra_rows > 0 {
+                        // we need to extend the mask here in case it's too short
+                        mask.extend(std::iter::repeat(true).take(extra_rows));
+                    }
+                    let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
+                    Ok(filtered_batch)
+                } else {
+                    Ok(record_batch)
+                }
+            }),
+            result_schema,
+        );
+        Ok(PyArrowType(Box::new(record_batch_iter)))
+    }
+}
+
+#[pyclass]
+struct PythonInterface(Box<dyn Engine + Send>);
+
+#[pymethods]
+impl PythonInterface {
+    #[new]
+    fn new(location: &str) -> DeltaPyResult<Self> {
+        let url = Url::parse(location).map_err(delta_kernel::Error::InvalidUrl)?;
+        let client = DefaultEngine::try_new(
+            &url,
+            HashMap::<String, String>::new(),
+            Arc::new(TokioBackgroundExecutor::new()),
+        )?;
+        Ok(PythonInterface(Box::new(client)))
+    }
+}
+
+/// Define the delta_kernel_python module. The name of this function _must_ be
+/// `delta_kernel_python`, and _must_ the `lib.name` setting in the `Cargo.toml`, otherwise Python
+/// will not be able to import the module.
+#[pymodule]
+fn delta_kernel_python(m: &Bound<'_, PyModule>) -> PyResult<()> {
+    m.add_class::<Table>()?;
+    m.add_class::<PythonInterface>()?;
+    m.add_class::<Snapshot>()?;
+    m.add_class::<ScanBuilder>()?;
+    m.add_class::<Scan>()?;
+    Ok(())
+}
```

Comment:
nit: new line at the end of file.

Also I'll skip the review of the rs code.

Reply:
fixed